### PR TITLE
fix(ReserveCurrencyLiquidator): improve how the contract handles reserve OR collateral currency shortfall

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

After testing out some more edge cases with the `ReserveCurrencyLiquidator` I realized there is a scenario that the contracts were not designed to handle and were not tested against. In particular what happens when the `ReserveCurrencyLiquidator` is told to liquidate a position that is larger than the sum of all reserve and collateral currency the DSProxy has access to? 

In the previous implementation, the transaction would revert. After this PR the DSProxy will liquidate as much as it can, using all reserves in the process.

**Summary**

Added extra logic and unit tests to the `ReserveCurrencyLiquidator` to accommodate a token shortfall.

**Details**

In addition to the make changes done in this PR I also refactored `ReserveCurrencyLiquidator` a small amount by pulling out common logic into some functions that can be re-used.

I also realized that the tests that were meant to test this within the `Liquidator` Tests were wrong. After this PR this has been addressed.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/liquidator/test/Liquidator.js [hunk 1]
```diff
@@ -2028,6 +2028,22 @@ contract("Liquidator.js", function(accounts) {
             "Correctly deals with reserve being the same as collateral currency using DSProxy",
             async function() {
               // create a new liquidator and set the reserve currency to the collateral currency.
+              const proxyTransactionWrapper = new ProxyTransactionWrapper({
```
Comment (chrismaree) [inline packages/liquidator/test/Liquidator.js:2031]: this test was somewhat working correctly before but in an unintended way. effectively the `proxyTransactionWrapper` was not correctly initialized and was still using a reserve currency( not the collateral currency). The test below _does_ validate that shortfall behavior works as expected but in this test it was shortfalling on reserve (not also collateral). 

After fixing this it correctly values a shortfall in _both_ collateral and reserve.

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol [hunk 2]
```diff
@@ -27,6 +27,8 @@ contract ReserveCurrencyLiquidator {
      * passed liveness. At this point the position can be manually unwound.
      * @dev Any synthetics & collateral that the DSProxy already has are considered in the amount swapped and minted.
      * These existing tokens will be used first before any swaps or mints are done.
+     * @dev If there is a token shortfall (either from not enough reserve to buy sufficient collateral or not enough
+     * collateral to begins with or due to slippage) the script will liquidate as much as posable given the reserves.
```
Comment (nicholaspai) [inline packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:31]:
```suggestion
     * collateral to begins with or due to slippage) the script will liquidate as much as possible given the reserves.
```

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol [hunk 3]
```diff
@@ -52,32 +54,25 @@ contract ReserveCurrencyLiquidator {
...
-                FixedPoint.Unsigned(IERC20(fc.collateralCurrency()).balanceOf(address(this)))
-            );
+            subOrZero(totalCollateralRequired, getCollateralBalance(fc));
 
-        // 4.a. If there is some collateral to be purchased, execute a trade on uniswap to meet the shortfall.
+        // 4.b. If there is some collateral to be purchased, execute a trade on uniswap to meet the shortfall.
         // Note the path assumes a direct route from the reserve currency to the collateral currency.
-        // Note that if the reserve currency is equal to the collateral currency no trade will execute within the router.
-        if (collateralToBePurchased.isGreaterThan(0)) {
+        if (collateralToBePurchased.isGreaterThan(0) && reserveCurrency != fc.collateralCurrency()) {
```
Comment (nicholaspai) [inline packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:75]: +1

**Review by nicholaspai — APPROVED** [review]
LGTM

**Review by mrice32 — APPROVED** [review]
Looks great! I do think we should eventually refactor this monster function into chunkier helper functions, but this logic change LGTM!

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol [hunk 4]
```diff
@@ -92,21 +87,34 @@ contract ReserveCurrencyLiquidator {
                 deadline
             );
         }
+
+        // 4.c. If at this point we were not able to get the required amount of collateral (due to insufficient reserve
+        // or not enough collateral in the contract) the script should try to liquidate as much as it can regardless.
+        // Update the values of total collateral to the current collateral balance and re-compute the tokenShortfall
+        // as the maximum tokens that could be liquidated at the current GCR.
+        if (totalCollateralRequired.isGreaterThan(getCollateralBalance(fc))) {
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:95]: yup, it's all due to call stack limits being hit with internal variables.

**Pull request merged**